### ISSUE TITLE
Document cleaning up websocket resources

### ISSUE THREAD

**Issue by mvkra** [SuaveIO/suave#463]
# Forced disconnects from the websocket connection don't allow cleanup of resources

Using the example located at https://github.com/SuaveIO/suave/issues/307 I'm having issues handling when a client disconnects forcibly (doesn't send a Close message). I want to dispose of resources held by the connection at this point (effectively run the Close logic).

Reproducing it is easy to do: Start an FSI session, connect to a Suave webserver and reset the FSI client session. Is there a way I can code the socket monad or make my own so that it handles this scenario? Thanks.

```
socket {
            while !loop do
                let! m = ws.read()
                match m with
                | Text,data,true -> () //process incoming message
                | Ping,_,_ -> 
                    inbox.Post (Pong, [||], true)
                | Close,_,_ -> 
                    // This doesn't fire when the FSI session is closed
                    inbox.Post (Close, [||], true)
                    loop := false
                | _ -> ()
        }
```

**Comment by haf** [comment]
@mvkra Thank you for reporting. It's a known limitation right now, so if you're affected by it, we'd be very happy to help you out with code reviewing a PR of your making.

**Comment by mvkra** [comment]
I thought it may of been my misunderstanding. As a result I came up with a solution that works well but isn't so intuitive for a new user to do (due to the fact that I don't know how to in most editors to see what the actual type of Socket<'a> is) - bascially an async expression that wraps around the socket expression like so. See below for any other people experiencing my issue:

```
/// Handles any errors overall to the connection
async {
    let! socketHandlerWithPotentialErrors = suaveReceiveSocketHandler // The socket computational expression
    match socketHandlerWithPotentialErrors with
    | Choice1Of2() -> ()
    | Choice2Of2(err) ->
        sprintf "Error received on connection [Error: %A]" err
        // Do cleanup stuff
    return socketHandlerWithPotentialErrors
    }
```
From a design point of view we have a number of options:

1. Provide a function in the appropriate namespace to add a callback on error to the socket monad and return another socket monad: Socket<'a> -> (Error -> unit) -> Socket<'a>
2.  Add an argument to the "socket" computation expression builder with this callback. Any exit of the monad will call this callback with either Success or Error cases. This forces you to provide it which with most websocket connections I've seen is probably a good idea. However it is a breaking change.
3. Any other ideas/thoughts?

**Comment by ademar** [comment]
Wouldn't a pattern `use IDisposable` work in your case ?

**Comment by mvkra** [comment]
A simple use statement inside the socket monad would work if it triggered upon the user disconnecting - my testing showed it didn't call any of the disposable logic when doing this. In our case I wouldn't feel comfortable using it unless I had the option of capturing the error and log it with some other detail. While this is possible as per my code above it isn't obvious to a newcomer. Obviously an IDisposable doesn't catch errors like the async handler I showed above does. Since I've got a workaround I'm happy to close this issue if your happy with the way it is.
